ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Eras.hs

```diff
@@ -132,23 +147,162 @@ class ( SL.ShelleyBasedEra era
+
+-- | The default implementation of 'applyShelleyBasedTx', a thin wrapper around
+-- 'SL.applyTx'
+defaultApplyShelleyBasedTx :: forall era.
+     ShelleyBasedEra era
+  => SL.Globals
+  -> SL.LedgerEnv    era
+  -> SL.MempoolState era
+  -> WhetherToIntervene
+  -> SL.Tx           era
+  -> Except
+       (ShelleyApplyTxError era)
+       ( SL.MempoolState era
+       , TxInBlock       era
+       )
+defaultApplyShelleyBasedTx globals ledgerEnv mempoolState _wti tx =
+      withExcept (toShelleyApplyTxError (Proxy @era))
+    $ SL.applyTx
+        globals
+        ledgerEnv
+        mempoolState
+        tx
+
+instance (SL.PraosCrypto c, ShowProxy (SL.ApplyTxError (ShelleyEra c))) => ShelleyBasedEra (ShelleyEra c) where
+  type ShelleyApplyTxError (ShelleyEra c) = SL.ApplyTxError (ShelleyEra c)
+
   shelleyBasedEraName _ = "Shelley"
 
-instance SL.PraosCrypto c => ShelleyBasedEra (AllegraEra c) where
+  toShelleyApplyTxError _prx = id
+
+  applyShelleyBasedTx = defaultApplyShelleyBasedTx
+
+instance (SL.PraosCrypto c, ShowProxy (SL.ApplyTxError (AllegraEra c))) => ShelleyBasedEra (AllegraEra c) where
+  type ShelleyApplyTxError (AllegraEra c) = SL.ApplyTxError (AllegraEra c)
+
   shelleyBasedEraName _ = "Allegra"
 
-instance SL.PraosCrypto c => ShelleyBasedEra (MaryEra c) where
+  toShelleyApplyTxError _prx = id
+
+  applyShelleyBasedTx = defaultApplyShelleyBasedTx
+
+instance (SL.PraosCrypto c, ShowProxy (SL.ApplyTxError (MaryEra c))) => ShelleyBasedEra (MaryEra c) where
+  type ShelleyApplyTxError (MaryEra c) = SL.ApplyTxError (MaryEra c)
+
   shelleyBasedEraName _ = "Mary"
 
+  toShelleyApplyTxError _prx = id
+
+  applyShelleyBasedTx = defaultApplyShelleyBasedTx
+
 instance SL.PraosCrypto c => ShelleyBasedEra (AlonzoEra c) where
+  type ShelleyApplyTxError (AlonzoEra c) = AlonzoApplyTxError c
+
   shelleyBasedEraName _ = "Alonzo"
 
+  toShelleyApplyTxError _prx = AlonzoApplyTxErrorProper
+
+  applyShelleyBasedTx globals ledgerEnv mempoolState wti tx = do
+      (mempoolState', vtx) <- defaultApplyShelleyBasedTx
+        globals
+        ledgerEnv
+        mempoolState
+        wti
+        tx
+
+      let Alonzo.IsValidating scriptsOK = Alonzo.isValidating vtx
+      case wti of
+        Intervene | not scriptsOK ->
+            throwError
+          $ AlonzoApplyTxErrorPromotedWarning
+          $ AlonzoPlutusErrors []
+        _ -> pure ()
+
+      pure (mempoolState', vtx)
+
+-- | An /error/ arising from applying an Alonzo transaction
+data AlonzoApplyTxError c
+  -- | The transaction is invalid
+  = AlonzoApplyTxErrorProper          !(SL.ApplyTxError (AlonzoEra c))
+  -- | The transaction is valid, but it has warnings that were promoted to
+  -- errors as part of an intervention
+  --
+  -- See 'WhetherToIntervene'.
+  | AlonzoApplyTxErrorPromotedWarning !AlonzoApplyTxWarning
+  deriving (Eq, Show)
+
+instance Typeable c => ShowProxy (AlonzoApplyTxError c) where
+
+instance SL.PraosCrypto c => ToCBOR (AlonzoApplyTxError c) where
```

Comment:
This serialization is not backwards compatible. However, this is only used for the new `AlonzoEra` so it need not be backwards compatible.

I do wonder if this is too clever/too subtle. Something like: future eras will probably have this same structure, so should the Consensus API reflect it (... maybe adding `WhetherToIntervene` is enough?), or should it remain "hidden" structure inside of integration code like this?

Reply:
Outdated -- PR input-output-hk/cardano-ledger-specs#2379 supplants this